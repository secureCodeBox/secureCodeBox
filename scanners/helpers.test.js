const { scan } = require('./helpers');

jest.mock('@kubernetes/client-node', () => {
  // Lazily require the actual module to be able to use variables like CustomObjectsApi
  const actualKubernetesClientNode = jest.requireActual('@kubernetes/client-node');

  const { CustomObjectsApi, CoreV1Api, BatchV1Api } = actualKubernetesClientNode;

  const mockKubeConfigInstance = {
    makeApiClient: jest.fn((apiType) => {
      switch (apiType) {
        case CustomObjectsApi:
          return {
            createNamespacedCustomObject: jest.fn(),
            deleteNamespacedCustomObject: jest.fn(),
            getNamespacedCustomObjectStatus: jest.fn(),
            listNamespacedCustomObject: jest.fn(),
          };
        case CoreV1Api:
          return {
            readNamespacedPodLog: jest.fn(),
            listNamespacedPod: jest.fn(),
          };
        case BatchV1Api:
          return {
            listNamespacedJob: jest.fn(),
          };
        default:
          throw new Error('API type not supported');
      }
    }),
  };

  return {
    KubeConfig: jest.fn(() => mockKubeConfigInstance),
    ...actualKubernetesClientNode, // Spread the actual module exports
  };
});

describe('helpers', () => {
  let mockCreateNamespacedCustomObject, mockDeleteNamespacedCustomObject, mockGetNamespacedCustomObjectStatus,
      mockReadNamespacedPodLog, mockListNamespacedPod, mockListNamespacedJob, mockListNamespacedCustomObject;

  const { KubeConfig, CustomObjectsApi, CoreV1Api, BatchV1Api } = require('@kubernetes/client-node');

  beforeAll(() => {
    // Mock the KubeConfig constructor TODO: This is not working
    const kc = new KubeConfig(); 

    mockCreateNamespacedCustomObject = jest.fn();
    mockDeleteNamespacedCustomObject = jest.fn();
    mockGetNamespacedCustomObjectStatus = jest.fn();
    mockReadNamespacedPodLog = jest.fn();
    mockListNamespacedPod = jest.fn();
    mockListNamespacedJob = jest.fn();
    mockListNamespacedCustomObject = jest.fn();

    kc.makeApiClient = jest.fn().mockImplementation((apiType) => {
      if (apiType === CustomObjectsApi) {
        return {
          createNamespacedCustomObject: mockCreateNamespacedCustomObject,
          deleteNamespacedCustomObject: mockDeleteNamespacedCustomObject,
          getNamespacedCustomObjectStatus: mockGetNamespacedCustomObjectStatus,
          listNamespacedCustomObject: mockListNamespacedCustomObject,
        };
      } else if (apiType === CoreV1Api) {
        return {
          readNamespacedPodLog: mockReadNamespacedPodLog,
          listNamespacedPod: mockListNamespacedPod,
        };
      } else if (apiType === BatchV1Api) {
        return {
          listNamespacedJob: mockListNamespacedJob,
        };
      }
      throw new Error('API type not supported');
    });
  });

});
