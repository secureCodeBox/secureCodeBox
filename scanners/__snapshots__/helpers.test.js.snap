// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kubernetes interaction tests cascading scan function should create a cascading scan and return findings on successful completion 1`] = `
{
  "categories": {
    "Discovered Credentials": 1,
  },
  "count": 1,
  "severities": {
    "high": 1,
  },
}
`;

exports[`Kubernetes interaction tests scan function should create a scan and return findings on successful completion 1`] = `
{
  "categories": {
    "Vulnerability": 24,
  },
  "count": 24,
  "severities": {
    "high": 24,
  },
}
`;

exports[`Kubernetes interaction tests scan function should create a scan and return findings on successful completion 2`] = `
[MockFunction] {
  "calls": [
    [
      "execution.securecodebox.io",
      "v1",
      "integration-tests",
      "scans",
      {
        "apiVersion": "execution.securecodebox.io/v1",
        "kind": "Scan",
        "metadata": {
          "generateName": "typo3scan-old-typo3-",
        },
        "spec": {
          "initContainers": [],
          "parameters": [],
          "scanType": "typo3scan",
          "volumeMounts": [],
          "volumes": [],
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Kubernetes interaction tests scan function should create a scan and return findings on successful completion 3`] = `
[MockFunction] {
  "calls": [
    [
      "execution.securecodebox.io",
      "v1",
      "integration-tests",
      "scans",
      "typo3scan-old-typo3-pw8vt",
    ],
    [
      "execution.securecodebox.io",
      "v1",
      "integration-tests",
      "scans",
      "typo3scan-old-typo3-pw8vt",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
